### ISSUE TITLE
`diff-hl-mode` is invisible in terminal Emacs (needs `diff-hl-margin-mode`)

### ISSUE THREAD

**Issue by jessy836** [doomemacs/doomemacs#8001]
### I confirm that

- [x] I have searched the [issue tracker](https://github.com/doomemacs/doomemacs/issues), [documentation](https://docs.doomemacs.org), [FAQ](https://docs.doomemacs.org/-/faq), [Discourse](https://discourse.doomemacs.org), and [Google](https://google.com), in case this issue has already been reported/resolved.
- [x] I have read "[How to Debug Issues](https://doomemacs.org/d/how2debug)", and will use it to provide as much information about this issue as possible.
- [x] The issue can be reproduced on the **latest** available commit of Doom.
- [x] The issue can be reproduced on a stable release of Emacs, such as 27, 28, or 29. *(Unstable versions end in .50, .60, or .9x)*

### Expected behavior

when emacs is open in terminal:
- in dired/dirvish having a indicator when a file is unstage
- in a buffer having an indicator which indicate unstaged lines
- maybe o  other indicator  I'm not aware 

### Current behavior

By default no indicator   is displayed

When I execute: 

M-x diff-hl-margin-mode

I get the indicator

It will be possible to add a condition to use diff-hl-margin-mode by default when :os tty is enable ?

### Steps to reproduce

In a directory under git control and containing an unstaged file:

1. emacs -nw .

### System Information

https://pastebin.com/EbFM92ig

**Comment by hlissner** [comment]
> It will be possible to add a condition to use diff-hl-margin-mode by default when :os tty is enable ?

A better solution is needed. There are many users who use both TTY and GUI Emacs (some within the same daemon session), so enabling `diff-hl-margin-mode` based on `:os tty` alone casts too wide a net. 

Perhaps I'll cycle `global-diff-hl-mode` and `diff-hl-margin-mode` when the user focuses a terminal frame after a GUI frame. See if this works for you:

```elisp
;;; add to $DOOMDIR/config.el
(add-hook! 'doom-switch-frame-hook
  (defun adjust-diff-hl-in-tty-emacs-h ()
    (let ((this-frame-type (if (display-graphic-p) 'gui 'tty)))
      (unless (eq (get 'diff-hl-mode 'last) this-frame-type)
        (global-diff-hl-mode -1)
        (diff-hl-margin-mode (if (display-graphic-p) -1 +1))
        (global-diff-hl-mode +1)
        (put 'diff-hl-mode 'last this-frame-type)))))
```

**Comment by jessy836** [comment]
> 
> ```elisp
> ;;; add to $DOOMDIR/config.el
> (add-hook! 'doom-switch-frame-hook
>   (defun adjust-diff-hl-in-tty-emacs-h ()
>     (let ((this-frame-type (if (display-graphic-p) 'gui 'tty)))
>       (unless (eq (get 'diff-hl-mode 'last) this-frame-type)
>         (global-diff-hl-mode -1)
>         (diff-hl-margin-mode (if (display-graphic-p) -1 +1))
>         (global-diff-hl-mode +1)
>         (put 'diff-hl-mode 'last this-frame-type)))))
> ```

Thank, I tryed to put that in My  config.el and then open emacs -nw in a  directory containing unstaged file but it changed nothing. I yet had to manually activate diff-hl-margin-mode

**Comment by jessy836** [comment]
I tryed something to understand why the workaround suggested by @hlissner doesn't works. I want to know if `doom-switch-frame-hook` is  triggered. I added the following in my init.el

``` emacs-lisp
(setq jessy836-count 0)
(add-hook! 'doom-switch-frame-hook
(defun jessy836-xp ()
(setq jessy836-count (1+ jessy836-count))))
```

Then I open emacs and dired/dirvish with `spc o -`  . The value of `jessy836-count`  has remained at 0.
The workaround I've found for the moment is to use `doom-switch-window-hook` instead of `doom-switch-frame-hook` .
